backends/p4tools/common/lib/trace_event_types.cpp

```diff
@@ -50,15 +50,28 @@ const Expression *Expression::apply(Transform &visitor) const {
 }
 
 const Expression *Expression::evaluate(const Model &model, bool doComplete) const {
-    if (Taint::hasTaint(value)) {
-        return new Expression(&Taint::TAINTED_STRING_LITERAL, label);
+    const IR::Expression *evaluatedValue = nullptr;
+    if (const auto *structExpr = value->to<IR::StructExpression>()) {
+        evaluatedValue = model.evaluateStructExpr(structExpr, doComplete);
+    } else if (const auto *listExpr = value->to<IR::BaseListExpression>()) {
+        evaluatedValue = model.evaluateListExpr(listExpr, doComplete);
```

Comment:
I'm curious, why this has to be done here and not in the model itself?

Reply:
The problem is simply typing. `evaluate` returns an `IR::Literal`. The expectation of calling evaluate is that it will give you a `IR::Literal` or a list/struct of literals. 

To simplify I would need to add another function that returns `IR::Expression` to the model or simplify the entire thing to return `IR::Expression`.

Reply:
I see, OK. It is probably better to keep the more concrete types unless we need to go over all the options in many places.